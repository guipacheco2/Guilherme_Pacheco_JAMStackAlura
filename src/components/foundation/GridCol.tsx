import React from 'react'
import styled, {
  css,
  CSSProperties,
  FlattenSimpleInterpolation,
} from 'styled-components'
import {
  breakpointsMedia,
  createBreakpoints,
  propsToStyle,
  ResponsiveBreakpoints,
} from '../../theme'

export interface StyledGridColProps {
  size?: ResponsiveBreakpoints<number>
  offset?: ResponsiveBreakpoints<number>
  alignItems?: ResponsiveBreakpoints<CSSProperties['alignItems']>
  justifyContent?: ResponsiveBreakpoints<CSSProperties['justifyContent']>
  flexDirection?: ResponsiveBreakpoints<CSSProperties['flexDirection']>
  marginTop?: ResponsiveBreakpoints<CSSProperties['marginTop']>
}

const StyledGridCol = styled.div<StyledGridColProps>(
  ({ alignItems, justifyContent, flexDirection, marginTop }) => {
    return css`
      padding-right: 16px;
      padding-left: 16px;
      flex-basis: 0;
      flex-grow: 1;
      max-width: 100%;
      display: flex;
      ${gridSize}
      ${gridOffset}
      ${propsToStyle({
        alignItems,
        justifyContent,
        flexDirection,
        marginTop,
      })}
    `
  },
)

interface GridColProps extends StyledGridColProps {
  children: React.ReactNode
}

export function GridCol({
  alignItems,
  justifyContent,
  flexDirection,
  marginTop,
  children,
  size,
  offset,
}: GridColProps): JSX.Element {
  return (
    <StyledGridCol
      alignItems={alignItems}
      size={size}
      offset={offset}
      justifyContent={justifyContent}
      flexDirection={flexDirection}
      marginTop={marginTop}
    >
      {children}
    </StyledGridCol>
  )
}

function gridSize({ size }: StyledGridColProps): FlattenSimpleInterpolation {
  if (!size) return []

  if (typeof size === 'number') {
    return css`
      flex: 0 0 ${(100 * size) / 12}%;
      max-width: ${(100 * size) / 12}%;
    `
  }

  const breakpoints = createBreakpoints(
    size,
    (s) => css`
      flex: 0 0 ${(100 * s) / 12}%;
      max-width: ${(100 * s) / 12}%;
    `,
  )

  return breakpointsMedia(breakpoints)
}

function gridOffset({
  offset,
}: StyledGridColProps): FlattenSimpleInterpolation {
  if (!offset) return []

  if (typeof offset === 'number') {
    return css`
      margin-left: ${(100 * offset) / 12}%;
    `
  }

  const breakpoints = createBreakpoints(
    offset,
    (o) => css`
      margin-left: ${(100 * o) / 12}%;
    `,
  )

  return breakpointsMedia(breakpoints)
}
