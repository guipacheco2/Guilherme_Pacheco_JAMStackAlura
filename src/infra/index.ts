export * from './hooks'
export * from './test'
