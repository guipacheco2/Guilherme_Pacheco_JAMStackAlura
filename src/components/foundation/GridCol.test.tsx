import React from 'react'
import { render } from '../../infra'
import { GridCol } from './GridCol'

describe('<GridCol />', () => {
  describe('when size', () => {
    test('is empty', () => {
      const { baseElement } = render(<GridCol>child</GridCol>)

      expect(baseElement).toMatchSnapshot()
    })

    test('is number', () => {
      const { baseElement } = render(<GridCol size={12}>child</GridCol>)

      expect(baseElement).toMatchSnapshot()
    })

    test('is responsive', () => {
      const { baseElement } = render(<GridCol size={{ sm: 6 }}>child</GridCol>)

      expect(baseElement).toMatchSnapshot()
    })
  })

  describe('when offset', () => {
    test('is number', () => {
      const { baseElement } = render(<GridCol offset={1}>child</GridCol>)

      expect(baseElement).toMatchSnapshot()
    })

    test('is responsive', () => {
      const { baseElement } = render(
        <GridCol offset={{ sm: 3 }}>child</GridCol>,
      )

      expect(baseElement).toMatchSnapshot()
    })
  })
})
