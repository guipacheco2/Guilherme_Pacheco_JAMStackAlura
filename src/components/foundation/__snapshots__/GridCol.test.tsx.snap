// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<GridCol /> when offset is number 1`] = `
<body>
  .c0 {
  padding-right: 16px;
  padding-left: 16px;
  -webkit-flex-basis: 0;
  -ms-flex-preferred-size: 0;
  flex-basis: 0;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  max-width: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  margin-left: 8.333333333333334%;
}

<div>
    <div
      class="c0"
      offset="1"
    >
      child
    </div>
  </div>
</body>
`;

exports[`<GridCol /> when offset is responsive 1`] = `
<body>
  .c0 {
  padding-right: 16px;
  padding-left: 16px;
  -webkit-flex-basis: 0;
  -ms-flex-preferred-size: 0;
  flex-basis: 0;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  max-width: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

@media only screen and (min-width:600px) {
  .c0 {
    margin-left: 25%;
  }
}

<div>
    <div
      class="c0"
      offset="[object Object]"
    >
      child
    </div>
  </div>
</body>
`;

exports[`<GridCol /> when size is empty 1`] = `
.c0 {
  padding-right: 16px;
  padding-left: 16px;
  -webkit-flex-basis: 0;
  -ms-flex-preferred-size: 0;
  flex-basis: 0;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  max-width: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

<body>
  <div>
    <div
      class="c0"
    >
      child
    </div>
  </div>
</body>
`;

exports[`<GridCol /> when size is number 1`] = `
<body>
  .c0 {
  padding-right: 16px;
  padding-left: 16px;
  -webkit-flex-basis: 0;
  -ms-flex-preferred-size: 0;
  flex-basis: 0;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  max-width: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex: 0 0 100%;
  -ms-flex: 0 0 100%;
  flex: 0 0 100%;
  max-width: 100%;
}

<div>
    <div
      class="c0"
      size="12"
    >
      child
    </div>
  </div>
</body>
`;

exports[`<GridCol /> when size is responsive 1`] = `
<body>
  .c0 {
  padding-right: 16px;
  padding-left: 16px;
  -webkit-flex-basis: 0;
  -ms-flex-preferred-size: 0;
  flex-basis: 0;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  max-width: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

@media only screen and (min-width:600px) {
  .c0 {
    -webkit-flex: 0 0 50%;
    -ms-flex: 0 0 50%;
    flex: 0 0 50%;
    max-width: 50%;
  }
}

<div>
    <div
      class="c0"
    >
      child
    </div>
  </div>
</body>
`;
